### PR TITLE
Update the stub code of NDM service!

### PR DESCRIPTION
**Pull request by JamePeng**
"GetCurrentState", "GetTargetState" , I am so comfused, what is the mean "current" and "target"

### CODE REVIEW

File: src/core/hle/service/ndm/ndm.cpp [hunk 1]
```diff
@@ -11,28 +15,192 @@
...
 
     cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) enable_daemon_lock=0x%08X ", enable_daemon_lock);
+}
+
+void SuspendDaemons(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( (u32)default_daemon_bit_mask & (~bit_mask) );
+    for (u32 index = 0; index < ARRAY_SIZE(daemon_status); ++index) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
for (size_t index = 0; index < daemon_status.size(); ++index) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 2]
```diff
@@ -11,28 +15,192 @@
...
+        if (bit_mask & (0x00000001 << index)) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
if (bit_mask & (1 << index))
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 3]
```diff
@@ -11,28 +15,192 @@
...
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) daemon_bit_mask=0x%08X ", daemon_bit_mask);
 }
 
 void ResumeDaemons(Service::Interface* self) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( (u32)daemon_bit_mask | (bit_mask) );
+    for (u32 index = 0; index < ARRAY_SIZE(daemon_status); ++index) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
for (size_t index = 0; index < daemon_status.size(); ++index) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 4]
```diff
@@ -11,28 +15,192 @@
...
+        if (bit_mask & (0x00000001 << index)) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
if (bit_mask & (1 << index)) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 5]
```diff
@@ -11,28 +15,192 @@
...
+    cmd_buff[2] = retry_interval;
+    LOG_WARNING(Service_NDM, "(STUBBED) retry_interval=0x%08X ", retry_interval);
 }
 
 void OverrideDefaultDaemons(Service::Interface* self) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    default_daemon_bit_mask = static_cast<DaemonMask>(bit_mask);
+    daemon_bit_mask = default_daemon_bit_mask;
+    for (u32 index = 0; index < ARRAY_SIZE(daemon_status); ++index) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
for (size_t index = 0; index < daemon_status.size(); ++index) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 6]
```diff
@@ -11,28 +15,192 @@
...
+        if (bit_mask & (0x00000001 << index)) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
if (bit_mask & (1 << index)) {
```

File: src/core/hle/service/ndm/ndm.h [hunk 7]
```diff
@@ -12,36 +12,209 @@ class Interface;
...
+enum class Daemon : u32 {
+    Cec    = 0,
+    Boss   = 1,
+    Nim    = 2,
+    Friend = 3
+};
+
+enum class DaemonMask : u32 {
+    None    =  0,
+    Cec     = (1 << (u32)Daemon::Cec),
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.h]: static_cast

File: src/core/hle/service/ndm/ndm.cpp [hunk 8]
```diff
@@ -11,28 +15,192 @@
...
+    LOG_WARNING(Service_NDM, "(STUBBED) called");
+}
+
+void QueryStatus(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 daemon = cmd_buff[1] & 0xF;
 
-    LOG_WARNING(Service_NDM, "(STUBBED) bit_mask=0x%08X ", cmd_buff[1]);
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    cmd_buff[2] = static_cast<u32>(daemon_status[daemon]);
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
cmd_buff[2] = static_cast<u32>(daemon_status.at(daemon));
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 9]
```diff
@@ -11,28 +15,192 @@
...
+    enable_daemon_lock = false;
 
     cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) enable_daemon_lock=0x%08X ", enable_daemon_lock);
+}
+
+void SuspendDaemons(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( (u32)default_daemon_bit_mask & (~bit_mask) );
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: static_cast, also `(~bit_mask)` doesn't need to be parenthesized

File: src/core/hle/service/ndm/ndm.cpp [hunk 10]
```diff
@@ -11,28 +15,192 @@
...
+    }
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) daemon_bit_mask=0x%08X ", daemon_bit_mask);
 }
 
 void ResumeDaemons(Service::Interface* self) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( (u32)daemon_bit_mask | (bit_mask) );
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: static_cast, also `(bit_mask)` doesn't need to be parenthesized

File: src/core/hle/service/ndm/ndm.cpp [hunk 11]
```diff
@@ -2,6 +2,10 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
+#define DEFAULT_SCAN_INTERVAL 30
+#define DEFAULT_RETRY_INTERVAL 10
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: Don't use defines. Use concrete variables for these, or use a regular enum

File: src/core/hle/service/ndm/ndm.cpp [hunk 12]
```diff
@@ -11,28 +15,192 @@
 namespace Service {
 namespace NDM {
 
-void SuspendDaemons(Service::Interface* self) {
+static DaemonMask daemon_bit_mask = DaemonMask::Default;
+static DaemonMask default_daemon_bit_mask = DaemonMask::Default;
+static std::array<DaemonStatus, 4>daemon_status = { DaemonStatus::Idle, DaemonStatus::Idle, DaemonStatus::Idle, DaemonStatus::Idle };
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: space after the `>`

File: src/core/hle/service/ndm/ndm.cpp (hunk 11 above)
Comment (JamePeng) [inline src/core/hle/service/ndm/ndm.cpp]: @lioncash how about Use static const variables for these

Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: You can use static constexpr variables, or just use an enum like

```cpp
enum : u32 {
    DEFAULT_RETRY_INTERVAL = 10,
    DEFAULT_SCAN_INTERVAL  = 30,
};
```

since they're related in the sense they're both defaults.

Comment (JamePeng) [inline src/core/hle/service/ndm/ndm.cpp]: ok!

File: src/core/hle/service/ndm/ndm.cpp [hunk 13]
```diff
@@ -11,28 +12,192 @@
...
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) enable_daemon_lock=0x%08X ", enable_daemon_lock);
+}
+
+void SuspendDaemons(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( static_cast<u32>(default_daemon_bit_mask) & ~bit_mask );
+    for (size_t index = 0; index < ARRAY_SIZE(daemon_status); ++index) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
for (size_t index = 0; index < daemon_status.size(); ++index) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 14]
```diff
@@ -11,28 +12,192 @@
...
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) daemon_bit_mask=0x%08X ", daemon_bit_mask);
 }
 
 void ResumeDaemons(Service::Interface* self) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( static_cast<u32>(daemon_bit_mask) | bit_mask );
+    for (size_t index = 0; index < ARRAY_SIZE(daemon_status); ++index) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
for (size_t index = 0; index < daemon_status.size(); ++index) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 15]
```diff
@@ -11,28 +12,192 @@
...
+    cmd_buff[2] = static_cast<u32>(retry_interval);
+    LOG_WARNING(Service_NDM, "(STUBBED) retry_interval=0x%08X ", retry_interval);
 }
 
 void OverrideDefaultDaemons(Service::Interface* self) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    default_daemon_bit_mask = static_cast<DaemonMask>(bit_mask);
+    daemon_bit_mask = default_daemon_bit_mask;
+    for (size_t index = 0; index < ARRAY_SIZE(daemon_status); ++index) {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]:
```cpp
for (size_t index = 0; index < daemon_status.size(); ++index) {
```

File: src/core/hle/service/ndm/ndm.cpp [hunk 16]
```diff
@@ -11,28 +12,192 @@
...
+    enable_daemon_lock = false;
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_NDM, "(STUBBED) enable_daemon_lock=0x%08X ", enable_daemon_lock);
+}
+
+void SuspendDaemons(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 bit_mask = cmd_buff[1] & 0xF;
+    daemon_bit_mask = static_cast<DaemonMask>( static_cast<u32>(default_daemon_bit_mask) & ~bit_mask );
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: you don't need the spaces after `(` and before `)`

File: src/core/hle/service/ndm/ndm.h [hunk 17]
```diff
@@ -12,36 +12,214 @@ class Interface;
...
+
+enum class ExclusiveState : u32 {
+    None                = 0,
+    Infrastructure      = 1,
+    LocalCommunications = 2,
+    Streetpass          = 3,
+    StreetpassData      = 4,
+};
+
+enum class DefaultInterval : u32 {
```
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.h]: When I said make this an enum, I meant leave it in the .cpp file, but just do

```cpp
enum : u32 {
    DEFAULT_RETRY_INTERVAL = 10,
    DEFAULT_SCAN_INTERVAL  = 30
};
```

File: src/core/hle/service/ndm/ndm.cpp (hunk 11 above)
Comment (lioncash) [inline src/core/hle/service/ndm/ndm.cpp]: See [#discussion_r59024880](this comment).

File: src/core/hle/service/ndm/ndm.h (hunk 17 above)
Comment (JamePeng) [inline src/core/hle/service/ndm/ndm.h]: Actually , I had not use this enum constants before!
I got a new knowledge again!

Comment (lioncash) [inline src/core/hle/service/ndm/ndm.h]: This enum should be moved into the .cpp file, since it's an unscoped enum now.

Comment (JamePeng) [inline src/core/hle/service/ndm/ndm.h]: like this?
qq 20160408220714

Comment (lioncash) [inline src/core/hle/service/ndm/ndm.h]: Yes

File: src/core/hle/service/ndm/ndm.cpp [hunk 18]
```diff
@@ -11,28 +11,197 @@
...
+
+void GetRetryInterval(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    cmd_buff[2] = retry_interval;
+    LOG_WARNING(Service_NDM, "(STUBBED) retry_interval=0x%08X ", retry_interval);
 }
 
 void OverrideDefaultDaemons(Service::Interface* self) {
```
Comment (Subv) [inline src/core/hle/service/ndm/ndm.cpp]: What is this function supposed to do? What makes it different from ResumeDaemons?

File: src/core/hle/service/ndm/ndm.h [hunk 19]
```diff
@@ -12,36 +12,209 @@ class Interface;
...
+ *  Outputs:
+ *      1 : Result, 0 on success, otherwise error code
+ */
+void UnlockState(Service::Interface* self);
+
 /**
- *  SuspendDaemons
+ *  NDM::SuspendDaemons service function
  *  Inputs:
- *      0 : Command header (0x00020082)
```
Comment (Subv) [inline src/core/hle/service/ndm/ndm.h]: Why did you remove this line from all the documentation headers?

**Comment by Subv** [comment]
LGTM after my comments are addressed

File: src/core/hle/service/ndm/ndm.h (hunk 19 above)
Comment (JamePeng) [inline src/core/hle/service/ndm/ndm.h]: I don't think it is useful for the code!
otherwise, it is wrong addr when I looking at 3dbrew page
http://3dbrew.org/wiki/NDM_Services

**Comment by JamePeng** [comment]
I append the IPC::MakeHeader to function!

**Comment by bunnei** [comment]
LGTM
